File: android/app/src/main/java/com/lilru/liftr/ui/pets/OpponentPetChallengeSheet.kt
package com.lilru.liftr.ui.pets

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Button
import androidx.compose.material3.ButtonDefaults
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.ExperimentalMaterial3Api
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.ModalBottomSheet
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.material3.TextButton
import androidx.compose.material3.rememberModalBottomSheetState
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.platform.LocalContext
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import com.lilru.liftr.data.PetCombatChallengeMode
import com.lilru.liftr.data.PetCombatHeadToHeadSummaryWire
import com.lilru.liftr.data.PetCombatPetSummaryWire
import com.lilru.liftr.data.PetCombatPreviewWire
import com.lilru.liftr.prefs.LiftrPreferences
import com.lilru.liftr.ui.theme.liftrAppBackgroundGradient

@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun OpponentPetChallengeSheet(
    preview: PetCombatPreviewWire,
    defenderPet: PetCombatPetSummaryWire,
    headToHead: PetCombatHeadToHeadSummaryWire?,
    opponentUsername: String?,
    backgroundThemeId: String,
    onDismiss: () -> Unit,
    onChallenge: (Boolean) -> Unit,
    onRefreshPreview: suspend () -> PetCombatPreviewWire?,
    onPreviewRefreshed: (PetCombatPreviewWire) -> Unit,
    modifier: Modifier = Modifier
) {
    val context = LocalContext.current
    val sheetState = rememberModalBottomSheetState(skipPartiallyExpanded = true)
    var displayPreview by remember(preview) { mutableStateOf(preview) }
    var isRefreshingPreview by remember { mutableStateOf(false) }
    var showUnbalancedDialog by remember { mutableStateOf(false) }
    var dontShowAgain by remember { mutableStateOf(false) }
    var skipWarningState by remember { mutableStateOf(LiftrPreferences.skipPetCombatUnbalancedWarning(context)) }
    var selectedMode by remember {
        mutableStateOf(PetCombatChallengeMode.fromRaw(LiftrPreferences.petCombatChallengeMode(context)))
    }

    LaunchedEffect(Unit) {
        isRefreshingPreview = true
        try {
            onRefreshPreview()?.let { fresh ->
                displayPreview = fresh
                onPreviewRefreshed(fresh)
            }
        } finally {
            isRefreshingPreview = false
        }
    }

    fun resolvedSavedMode(): PetCombatChallengeMode =
        if (displayPreview.isAttackerUnderdog()) {
            PetCombatChallengeMode.fromRaw(LiftrPreferences.petCombatChallengeMode(context))
        } else {
            PetCombatChallengeMode.BALANCED
        }

    fun proceedToChallenge(disableNerfChoice: Boolean) {
        onDismiss()
        onChallenge(disableNerfChoice)
    }

    fun handleChallengeTap() {
        if (!displayPreview.isStatUnbalanced()) {
            proceedToChallenge(false)
            return
        }
        if (skipWarningState && !displayPreview.isAttackerUnderdog()) {
            proceedToChallenge(false)
            return
        }
        selectedMode = resolvedSavedMode()
        dontShowAgain = false
        showUnbalancedDialog = true
    }

    fun confirmChallenge() {
        if (dontShowAgain && !displayPreview.isAttackerUnderdog()) {
            LiftrPreferences.setSkipPetCombatUnbalancedWarning(context, true)
            skipWarningState = true
        }
        LiftrPreferences.setPetCombatChallengeMode(context, selectedMode.rawValue)
        val disableNerf = displayPreview.isAttackerUnderdog() && selectedMode.disableNerfChoice
        showUnbalancedDialog = false
        proceedToChallenge(disableNerf)
    }

    Box(modifier = modifier) {
        ModalBottomSheet(
            onDismissRequest = onDismiss,
            sheetState = sheetState,
            containerColor = Color.Transparent,
            dragHandle = null
        ) {
        Column(
            modifier = Modifier
                .fillMaxSize()
                .liftrAppBackgroundGradient(backgroundThemeId)
                .verticalScroll(rememberScrollState())
                .padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(12.dp)
        ) {
            Row(modifier = Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.End) {
                TextButton(onClick = onDismiss) { Text("Close") }
            }

            Column(
                modifier = Modifier.fillMaxWidth(),
                horizontalAlignment = Alignment.CenterHorizontally,
                verticalArrangement = Arrangement.spacedBy(8.dp)
            ) {
                PetCombatSummaryImage(pet = defenderPet, height = 120.dp)
                Text(defenderPet.displayName(), style = MaterialTheme.typography.titleMedium, fontWeight = FontWeight.Bold)
                PetRarityBadge(defenderPet.rarity)
            }

            if (headToHead != null && headToHead.totalBattles > 0) {
                PetCombatHeadToHeadSummaryView(
                    summary = headToHead,
                    opponentUsername = opponentUsername
                )
            }

            val attacker = displayPreview.attacker
            val defender = displayPreview.defender
            if (attacker?.pet != null && defender?.pet != null && attacker.stats != null && defender.stats != null) {
                PetCombatComparisonStats(
                    attackerName = attacker.pet.displayName(),
                    defenderName = defender.pet.displayName(),
                    attackerLevel = attacker.pet.currentLevel,
                    defenderLevel = defender.pet.currentLevel,
                    attackerStats = attacker.stats,
                    defenderStats = defender.stats
                )
            }

            displayPreview.energy?.let { energy ->
                Surface(
                    modifier = Modifier.fillMaxWidth(),
                    shape = RoundedCornerShape(12.dp),
                    color = Color.White.copy(alpha = 0.06f)
                ) {
                    Row(
                        modifier = Modifier.padding(12.dp),
                        horizontalArrangement = Arrangement.SpaceBetween,
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        PetEnergyBadge(energy = energy, label = "Your energy")
                        Text(
                            "Challenging costs 1 energy",
                            style = MaterialTheme.typography.labelSmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant
                        )
                    }
                }
            }

            when {
                isRefreshingPreview -> {
                    Box(
                        modifier = Modifier
                            .fillMaxWidth()
                            .padding(vertical = 8.dp),
                        contentAlignment = Alignment.Center
                    ) {
                        CircularProgressIndicator()
                    }
                }
                displayPreview.effectiveCanChallenge() -> {
                    Button(
                        onClick = { handleChallengeTap() },
                        modifier = Modifier.fillMaxWidth(),
                        colors = ButtonDefaults.buttonColors(containerColor = Color(0xFFFF9800))
                    ) {
                        Text("Challenge Pet", fontWeight = FontWeight.SemiBold)
                    }
                }
                else -> {
                    displayPreview.blockReasonText()?.let {
                        Text(
                            it,
                            style = MaterialTheme.typography.bodySmall,
                            color = MaterialTheme.colorScheme.onSurfaceVariant,
                            modifier = Modifier.fillMaxWidth(),
                            textAlign = TextAlign.Center
                        )
                    }
                }
            }

            Spacer(modifier = Modifier.padding(bottom = 24.dp))
        }
        }

        if (showUnbalancedDialog) {
            PetCombatUnbalancedMatchDialog(
                isUnderdog = displayPreview.isAttackerUnderdog(),
                selectedMode = selectedMode,
                onModeChange = { selectedMode = it },
                hardcoreBonusLabel = displayPreview.hardcoreBonusLabel(),
                dontShowAgain = dontShowAgain,
                onDontShowAgainChange = { dontShowAgain = it },
                onDismiss = { showUnbalancedDialog = false },
                onFight = { confirmChallenge() }
            )
        }
    }
}

File: Liftr/Pet/OpponentPetChallengeSheet.swift
import SwiftUI

struct OpponentPetChallengeSheet: View {
    @Environment(\.dismiss) private var dismiss
    @AppStorage("skipPetCombatUnbalancedWarning") private var skipPetCombatUnbalancedWarning = false
    @AppStorage("petCombatChallengeMode") private var savedChallengeModeRaw = PetCombatChallengeMode.balanced.rawValue

    let opponentUserId: UUID
    let preview: PetCombatPreview
    let defenderPet: PetCombatPetSummary
    let headToHead: PetCombatHeadToHeadSummary?
    let opponentUsername: String?
    let onChallenge: (Bool) -> Void
    let onPreviewRefreshed: (PetCombatPreview) -> Void

    @State private var displayPreview: PetCombatPreview
    @State private var isRefreshingPreview = false
    @State private var showUnbalancedDialog = false
    @State private var dontShowAgain = false
    @State private var selectedMode: PetCombatChallengeMode = .balanced

    init(
        opponentUserId: UUID,
        preview: PetCombatPreview,
        defenderPet: PetCombatPetSummary,
        headToHead: PetCombatHeadToHeadSummary?,
        opponentUsername: String?,
        onChallenge: @escaping (Bool) -> Void,
        onPreviewRefreshed: @escaping (PetCombatPreview) -> Void
    ) {
        self.opponentUserId = opponentUserId
        self.preview = preview
        self.defenderPet = defenderPet
        self.headToHead = headToHead
        self.opponentUsername = opponentUsername
        self.onChallenge = onChallenge
        self.onPreviewRefreshed = onPreviewRefreshed
        _displayPreview = State(initialValue: preview)
    }

    private var rarity: PetRarity {
        PetRarity(databaseValue: defenderPet.rarity) ?? .common
    }

    var body: some View {
        NavigationStack {
            ScrollView {
                VStack(spacing: 16) {
                    VStack(spacing: 10) {
                        PetCombatSummaryImage(pet: defenderPet, height: 120)

                        Text(defenderPet.displayName)
                            .font(.title3.bold())

                        PetRarityBadge(rarity: rarity)
                    }

                    if let headToHead, headToHead.totalBattles > 0 {
                        PetCombatHeadToHeadSummaryView(
                            summary: headToHead,
                            opponentUsername: opponentUsername
                        )
                    }

                    if let attacker = displayPreview.attacker,
                       let defender = displayPreview.defender,
                       let attackerPet = attacker.pet,
                       let defenderPet = defender.pet,
                       let attackerStats = attacker.stats,
                       let defenderStats = defender.stats {
                        PetCombatComparisonStats(
                            attackerName: attackerPet.displayName,
                            defenderName: defenderPet.displayName,
                            attackerStats: attackerStats,
                            defenderStats: defenderStats,
                            attackerLevel: attackerPet.currentLevel,
                            defenderLevel: defenderPet.currentLevel
                        )
                    }

                    if let energy = displayPreview.energy {
                        HStack {
                            PetEnergyBadge(energy: energy, label: "Your energy")
                            Spacer()
                            Text("Challenging costs 1 energy")
                                .font(.caption2)
                                .foregroundStyle(.secondary)
                        }
                        .padding(12)
                        .background(.white.opacity(0.06), in: RoundedRectangle(cornerRadius: 12, style: .continuous))
                    }

                    if isRefreshingPreview {
                        ProgressView()
                            .frame(maxWidth: .infinity)
                    } else if displayPreview.effectiveCanChallenge {
                        Button(action: handleChallengeTap) {
                            Label("Challenge Pet", systemImage: "bolt.horizontal.circle.fill")
                                .font(.subheadline.weight(.semibold))
                                .frame(maxWidth: .infinity)
                        }
                        .buttonStyle(.borderedProminent)
                        .tint(.orange)
                    } else if let reason = displayPreview.blockReasonText {
                        Text(reason)
                            .font(.footnote)
                            .foregroundStyle(.secondary)
                            .multilineTextAlignment(.center)
                            .frame(maxWidth: .infinity)
                    }
                }
                .frame(maxWidth: .infinity)
                .padding()
            }
            .navigationBarTitleDisplayMode(.inline)
            .toolbar {
                ToolbarItem(placement: .topBarTrailing) {
                    Button("Close") { dismiss() }
                }
            }
        }
        .gradientBG()
        .overlay {
            if showUnbalancedDialog {
                PetCombatUnbalancedMatchDialog(
                    isPresented: $showUnbalancedDialog,
                    dontShowAgain: $dontShowAgain,
                    isUnderdog: displayPreview.isAttackerUnderdog,
                    selectedMode: $selectedMode,
                    hardcoreBonusLabel: displayPreview.hardcoreBonusLabel,
                    onCancel: { showUnbalancedDialog = false },
                    onFight: confirmChallenge
                )
            }
        }
        .onAppear {
            selectedMode = PetCombatChallengeMode(rawValue: savedChallengeModeRaw) ?? .balanced
        }
        .task {
            await refreshPreview()
        }
    }

    private func refreshPreview() async {
        isRefreshingPreview = true
        defer { isRefreshingPreview = false }
        do {
            let fresh = try await PetService.shared.fetchCombatPreview(targetUserId: opponentUserId)
            displayPreview = fresh
            onPreviewRefreshed(fresh)
        } catch {
        }
    }

    private func handleChallengeTap() {
        if !displayPreview.isStatUnbalanced {
            dismiss()
            onChallenge(false)
            return
        }

        if skipPetCombatUnbalancedWarning && !displayPreview.isAttackerUnderdog {
            dismiss()
            onChallenge(false)
            return
        }

        selectedMode = resolvedSavedMode()
        dontShowAgain = false
        showUnbalancedDialog = true
    }

    private func confirmChallenge() {
        if dontShowAgain && !displayPreview.isAttackerUnderdog {
            skipPetCombatUnbalancedWarning = true
        }
        savedChallengeModeRaw = selectedMode.rawValue
        let disableNerf = displayPreview.isAttackerUnderdog && selectedMode.disableNerfChoice
        showUnbalancedDialog = false
        dismiss()
        onChallenge(disableNerf)
    }

    private func resolvedSavedMode() -> PetCombatChallengeMode {
        if displayPreview.isAttackerUnderdog {
            return PetCombatChallengeMode(rawValue: savedChallengeModeRaw) ?? .balanced
        }
        return .balanced
    }
}
